[DRAFT] https://issues.redhat.com/browse/ACM-11346 for 2.10 (Dev team still working on this) (#6811)
open-cluster-management/rhacm-docs: observability/use_observability.adoc

Use the observability service to view the utilization of clusters across your fleet.

//need to add required access

> 

Reviewer: Required access to do the below actions?

Reviewer: ```suggestion
Required access: Cluster administrator, the open-cluster-management:cluster-manager-admin role, or S3 administrator.
```

----

. Run the following command to add the access permission to you script file:

+
[source,bash]
----
chmod +x my_script.sh
----

. To verify updates to your script, run the following command to view the contents of your script:

+
[source,bash]
----
./my_script.sh $MY_TOKEN
----

+
Your script might resemble the following file:

+
----
#!/bin/bash -xe
TOKEN=${1:?}  # Your TOKEN
BASE_PATH="./tmp"
mkdir -p "${BASE_PATH}"

# ca.crt
crt=$(oc -n open-cluster-management-observability get secret observability-server-ca-certs -o json | jq -r '.data."ca.crt"' | base64 -d )
echo "${crt}" > "${BASE_PATH}"/ca.crt

PROXY_ROUTE_URL=$(oc get route rbac-query-proxy -n open-cluster-management-observability -o jsonpath='{.spec.host}')
QUERY_EXPRESSION="up"
curl -vv  --location --cacert "${BASE_PATH}"/ca.crt -H "Authorization: Bearer ${TOKEN}" https://${PROXY_ROUTE_URL}/api/v1/query\?query\=${QUERY_EXPRESSION}
----

. Get the default CA certificate and store the content of the key `tls.crt` in a local file. Run the following command:

+
[source,bash]
----
oc -n openshift-ingress get secret router-certs-default -o jsonpath="{.data.tls\.crt}" | base64 -d > ca.crt
----

> //not sure about this command^

Author: Thanks for your review @philipgough I have a few questions within the file. Is this command correct?

. To query metrics, run the following command:

+
[source,bash]
----
curl --cacert ./ca.crt -H "Authorization: Bearer {TOKEN}" https://{PROXY_ROUTE_URL}/api/v1/query?query={QUERY_EXPRESSION}
----
+
*Note:* The `QUERY_EXPRESSION` is the standard Prometheus query expression. For example, query the metrics `cluster_infrastructure_provider` by replacing the URL in the previously mentioned command with the following URL: `https://{PROXY_ROUTE_URL}/api/v1/query?query=cluster_infrastructure_provider`. For more details, see link:https://prometheus.io/docs/prometheus/latest/querying/basics/[Querying Prometheus].

> //are the steps on line 87-99 related to querying at all ? 

Author: Also here @philipgough

Reviewer 2: If so, we need to explain the reason why we are doing it and what problem we are addresing by doing it. What is the purpose of  proxy-byo-ca and  proxy-byo-cert? Also explain what are ca.crt and ca.key, ingress.crt and ingress.key, etc,

Reviewer: Afaics, the proxy-bring-your-own cert is because you're accessing an endpoint via MTLS that normally only authenticates with other ACM components via generated TLS certs. So the 'byo' cert is a cert the customer is providing to access the metrics directly.

Reviewer 2: @moadz can we explain that in the docs? Doesnt seem obvious to me at least. Also,shouldnt these byo-cert steps ordered first, just before queying the metrics step?

Author: @moadz

Use the observability service to view the utilization of clusters across your fleet.

*Required access:* Cluster administrator

- <<external-metric-query,Querying metrics using the observability API>>
- <<exporting-metrics-to-external-endpoints,Exporting metrics to external endpoints>>
- <<viewing-and-exploring-data,Viewing and exploring data>>

[#external-metric-query]
== Querying metrics using the observability API

To query metrics through the OpenShift `rbac-query-proxy` route, use the observability external API. Complete the following steps to get your queries for the `rbac-query-proxy` route:

. Get the details of the route with the following command:

+
[source,bash]
----
oc get route rbac-query-proxy -n open-cluster-management-observability
----

. You can also access the `rbac-query-proxy` route with your OpenShift OAuth access token. Run the following command to get the token that must be associated with a user or service account, which has permission to get namespaces:

+
[source,bash]
----
MY_TOKEN=$(oc whoami --show-token)
----

> . Run the following command to add the access permission to you script file:

Reviewer: ```suggestion
. Run the following command to add the access permission to your script file:
```

+
[source,bash]
----
chmod +x my_script.sh
----

. To verify updates to your script, run the following command to view the contents of your script:

+
[source,bash]
----
./my_script.sh $MY_TOKEN
----

> . Get the default CA certificate and store the content of the key `tls.crt` in a local file. Run the following command:

Author: Why does the user need to get the default CA certificate?

+
[source,bash]
----
oc -n openshift-ingress get secret router-certs-default -o jsonpath="{.data.tls\.crt}" | base64 -d > ca.crt
----

> . To query metrics, run the following command:

Reviewer: Worth clarifying that it's querying via direct API access - and not via the recommended grafana dashboard.

> . To query metrics, run the following command:

Author: suggestion based on conversation in Slack: https://redhat-internal.slack.com/archives/C07LGSSDNEM/p1725868053787269
```suggestion
. To access your endpoint by using the mutual TLS through the generated TLS certificates, run the following command to create `proxy-byo-ca` secrets:

+
[source,bash]
----
oc -n open-cluster-management-observability create secret tls proxy-byo-ca --cert ./ca.crt --key ./ca.key
----

. Create `proxy-byo-cert` secrets using the generated certificates by using the following command:

+
[source,bash]
----
oc -n open-cluster-management-observability create secret tls proxy-byo-cert --cert ./ingress.crt --key ./ingress.key
----

. To query metrics, run the following command:
```

+
[source,bash]
----
curl --cacert ./ca.crt -H "Authorization: Bearer {TOKEN}" https://{PROXY_ROUTE_URL}/api/v1/query?query={QUERY_EXPRESSION}
----
+
*Note:* The `QUERY_EXPRESSION` is the standard Prometheus query expression. For example, query the metrics `cluster_infrastructure_provider` by replacing the URL in the previously mentioned command with the following URL: `https://{PROXY_ROUTE_URL}/api/v1/query?query=cluster_infrastructure_provider`. For more details, see link:https://prometheus.io/docs/prometheus/latest/querying/basics/[Querying Prometheus].

> . To access your endpoint by using the mutual TLS through the generated TLS certificates, run the following command to create `proxy-byo-ca` secrets:

Reviewer: ```suggestion
. To access your endpoint by using mutual TLS through the generated TLS certificates, run the following command to create `proxy-byo-ca` secrets:
```

> . To access your endpoint by using the mutual TLS through the generated TLS certificates, run the following command to create `proxy-byo-ca` secrets:
> 
> +
> [source,bash]
> ----
> oc -n open-cluster-management-observability create secret tls proxy-byo-ca --cert ./ca.crt --key ./ca.key
> ----
> 
> . Create `proxy-byo-cert` secrets using the generated certificates by using the following command:
> 
> +
> [source,bash]
> ----
> oc -n open-cluster-management-observability create secret tls proxy-byo-cert --cert ./ingress.crt --key ./ingress.key
> ----

Author: Related to this suggestion: https://github.com/stolostron/rhacm-docs/pull/6811/files#r1750887326
```suggestion
```

Use the observability service to view the utilization of clusters across your fleet.

*Required access:* Cluster administrator

- <<external-metric-query,Querying metrics using the observability API>>
- <<exporting-metrics-to-external-endpoints,Exporting metrics to external endpoints>>
- <<viewing-and-exploring-data,Viewing and exploring data>>

> *Prerequsites:* Are there any prereqs that we need to list?

Author: @moadz

> *Prerequsites:* Are there any prereqs that we need to list?

Author: ```suggestion
[use-obs-prereq]
== Prerequsites

Are there any prereqs that we need to list?
```

Use the observability service to view the utilization of clusters across your fleet.

*Required access:* Cluster administrator

- <<external-metric-query,Querying metrics using the observability API>>
- <<exporting-metrics-to-external-endpoints,Exporting metrics to external endpoints>>
- <<viewing-and-exploring-data,Viewing and exploring data>>

> [use-obs-prereq]

Author: ```suggestion
[#use-obs-prereq]
```

== Prerequsites

> Are there any prereqs that we need to list?

Author: @moadz

[#external-metric-query]
== Querying metrics using the observability API

> To query metrics through the OpenShift `rbac-query-proxy` route, use the observability external API. Complete the following steps to get your queries for the `rbac-query-proxy` route:

Author: ```suggestion
To access your endpoint by using the mutual TLS, which verifies the identities of both parties in a network connection, query metrics through the OpenShift `rbac-query-proxy` route with the observability external API. Complete the following steps to get your queries for the `rbac-query-proxy` route:
```

. Get the details of the route with the following command:

+
[source,bash]
----
oc get route rbac-query-proxy -n open-cluster-management-observability
----

. You can also access the `rbac-query-proxy` route with your OpenShift OAuth access token. Run the following command to get the token that must be associated with a user or service account, which has permission to get namespaces:

+
[source,bash]
----
MY_TOKEN=$(oc whoami --show-token)
----

. Run the following command to add the access permission to your script file:

+
[source,bash]
----
chmod +x my_script.sh
----

. To verify updates to your script, run the following command to view the contents of your script:

+
[source,bash]
----

> ./my_script.sh $MY_TOKEN

Author: Remove the script steps from lines 39-53

> ----

Author: @jtudelag you made a comment about "a script that does not exist" in the Slack channel. I removed the script example because the doc team has moved away from documenting script examples. Since we are not providing the script, is your suggestion to remove this step? Alternatively, we can maybe add part of the script so that the user can see how the access permission might appear.

Reviewer: Yeah lets add extracts from the script, if we remove it we have to remove the whole section.

> . Get the default CA certificate and store the content of the `tls.crt` key in a local file. Run the following command:

Author: Why does the user need to get the default certificate?

Author: The reason is to access route

+
[source,bash]
----
oc -n openshift-ingress get secret router-certs-default -o jsonpath="{.data.tls\.crt}" | base64 -d > ca.crt
----

> . To access your endpoint by using the mutual TLS through the generated TLS certificates, run the following command to create `proxy-byo-ca` secrets:

Author: ```suggestion
. To access an endpoint of the default CA certificate from your hub cluster through mutual TLS, which verifies the identities of both parties in a network connection, create the `proxy-byo-ca` TLS secret with the following command:
```

+
[source,bash]
----
oc -n open-cluster-management-observability create secret tls proxy-byo-ca --cert ./ca.crt --key ./ca.key
----

> . Create `proxy-byo-cert` secrets using the generated certificates by using the following command:

Author: My attempt to explain the benefit of running this command
```suggestion
. To access an endpoint from your hub cluster through mutual TLS, create a `proxy-byo-cert` TLS secret using the generated `ingress.crt` certificate and `ingress.key` key. Run the following command:
```

+
[source,bash]
----
oc -n open-cluster-management-observability create secret tls proxy-byo-cert --cert ./ingress.crt --key ./ingress.key

> ----

Author: Request by @jtudelag to explain ingress.crt and ingress.key

[#exporting-metrics-to-external-endpoints]
== Exporting metrics to external endpoints

> //should this section be a subsection of the previous section, Querying metrics using the observability API?

Author: Wondering if Exporting metrics should be a subsection or is it fine that its the same header level as the previous section

Reviewer: I think it's probably fine as a subsection.

- For more information, see link:https://prometheus.io/docs/concepts/remote_write_spec/[Prometheus Remote-Write specification].

> - See link:https://docs.openshift.com/container-platform/4.13/authentication/managing-oauth-access-tokens.html[Managing user-owned OAuth access tokens].

Reviewer 2: Shuldnt we reference newer versions of OCP, rather than 4.13?

Reviewer 3: @jtudelag the doc links to the bottom part of OCP supported per n-2 strategy unless the product component was created ahead of that, so for 2.10 docs, we intentionally point to 4.13. This is because of testing and such and based on PM feedback and direction. I will make sure we make this clear in the doc, but for all other questions, we send users to the support matrix.

Reviewer 3: Mikela is working on clarifying this in install doc, as well as release notes.
